primitives/state-machine/src/trie_backend.rs

```diff
@@ -369,4 +591,320 @@ pub mod tests {
+					.with_optional_cache(cache.as_ref().map(|c| c.local_cache()))
+					.build();
+				assert_eq!(proving.storage(&[42]).unwrap().unwrap(), vec![42; size_content]);
+
+				let proof = proving.extract_proof().unwrap().unwrap();
+
+				let proof_check =
+					create_proof_check_backend::<BlakeTwo256>(in_memory_root.into(), proof)
+						.unwrap();
+				assert_eq!(proof_check.storage(&[42]).unwrap().unwrap(), vec![42; size_content]);
+			}
+		}
+	}
+
+	#[test]
+	fn proof_record_works_with_iter() {
+		proof_record_works_with_iter_inner(StateVersion::V0);
+		proof_record_works_with_iter_inner(StateVersion::V1);
+	}
+	fn proof_record_works_with_iter_inner(state_version: StateVersion) {
+		for cache in
+			[Some(SharedTrieCache::new(Configuration { maximum_size_in_bytes: 1024 * 1024 })), None]
+		{
+			// Run multiple times to have a filled cache.
+			for _ in 0..3 {
+				let contents = (0..64).map(|i| (vec![i], Some(vec![i]))).collect::<Vec<_>>();
+				let in_memory = InMemoryBackend::<BlakeTwo256>::default();
+				let in_memory = in_memory.update(vec![(None, contents)], state_version);
+				let in_memory_root = in_memory.storage_root(std::iter::empty(), state_version).0;
+				(0..64)
+					.for_each(|i| assert_eq!(in_memory.storage(&[i]).unwrap().unwrap(), vec![i]));
+
+				let trie = in_memory.as_trie_backend().unwrap();
+				let trie_root = trie.storage_root(std::iter::empty(), state_version).0;
+				assert_eq!(in_memory_root, trie_root);
+				(0..64).for_each(|i| assert_eq!(trie.storage(&[i]).unwrap().unwrap(), vec![i]));
+
+				let proving = TrieBackendBuilder::wrap(&trie)
+					.with_recorder(Recorder::default())
+					.with_optional_cache(cache.as_ref().map(|c| c.local_cache()))
+					.build();
+
+				(0..63).for_each(|i| {
+					assert_eq!(proving.next_storage_key(&[i]).unwrap(), Some(vec![i + 1]))
+				});
+
+				let proof = proving.extract_proof().unwrap().unwrap();
+
+				let proof_check =
+					create_proof_check_backend::<BlakeTwo256>(in_memory_root.into(), proof)
+						.unwrap();
+				(0..63).for_each(|i| {
+					assert_eq!(proof_check.next_storage_key(&[i]).unwrap(), Some(vec![i + 1]))
+				});
+			}
+		}
+	}
+
+	#[test]
+	fn proof_recorded_and_checked_with_child() {
+		proof_recorded_and_checked_with_child_inner(StateVersion::V0);
+		proof_recorded_and_checked_with_child_inner(StateVersion::V1);
+	}
+	fn proof_recorded_and_checked_with_child_inner(state_version: StateVersion) {
+		let child_info_1 = ChildInfo::new_default(b"sub1");
+		let child_info_2 = ChildInfo::new_default(b"sub2");
+		let child_info_1 = &child_info_1;
+		let child_info_2 = &child_info_2;
+		let contents = vec![
+			(None, (0..64).map(|i| (vec![i], Some(vec![i]))).collect::<Vec<_>>()),
+			(Some(child_info_1.clone()), (28..65).map(|i| (vec![i], Some(vec![i]))).collect()),
+			(Some(child_info_2.clone()), (10..15).map(|i| (vec![i], Some(vec![i]))).collect()),
+		];
+		let in_memory = InMemoryBackend::<BlakeTwo256>::default();
```

Comment:
Would be good to use a :

```
	pub type PrefixedInMemoryBackend<H> = TrieBackend<PrefixedMemoryDB<H>, H>;
```
may show the issue I foresee with child trie and rocksdb (but it may just not work for other reasons)

Reply:
Ahh that answers my other question.